UART Multidrop/Multipoint Support - Add base component, Dooya shades, Chenyang window openers, GM40 curtain motor

loongyh opened the pull request:
# What does this implement/fix? 

- Add UARTMulti base component
- Add Dooya Shade RS485 Motor [DM35EQL/J & DM45EQL/J](http://downloads.dooya.com/interiorapplication.pdf)
- Add Chenyang RS485 Window Opener [CP12, CP14, CP16, CP31, CP95](https://drive.google.com/file/d/16b1IaAimyyDuLjRlgnlAeW8OSLYSf_rX/view)
- Add [GM40](https://www.alibaba.com/product-detail/Galime-1-2m-Basic-Standard-Window_1600181919327.html) RS485 Curtain Motor

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (this will require users to update their yaml configuration files to keep working)

**Related issue or feature (if applicable):** Relevant but does not fix (this PR depends on RS485-TTL drivers with automatic flow control, i.e. no DE/RE control) https://github.com/esphome/feature-requests/issues/421 
fixes https://github.com/esphome/feature-requests/issues/1051

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1090
  
# Test Environment

- [x] ESP32
- [ ] ESP8266
- [x] Windows
- [ ] Mac OS
- [x] Linux

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml
uart:
  tx_pin: 17
  rx_pin: 16
  baud_rate: 9600
cover:
  - platform: dooya
    name: Dooya Shade
    address: 0xFEFE
    device_class: shade
  - platform: chenyang
    name: Chenyang Window
    address: 0xFF
    device_class: window
  - platform: gm40
    name: GM40 Curtain
    address: 0x00
    device_class: curtain
```

# Explain your changes

This PR adds a base `UARTMulti` 'router' component that forwards received bytes to registered UARTMulti components to support multidrop/multipoint devices from any manufacturer (as long as they have the same serial settings e.g. baud rate etc.). Inspired heavily from the modbus component.
This will add support for multipoint/multidrop buses such as RS485 over UART using a RS485-TTL converter.
Also adds in support for the RS485 Dooya blinds/shades motor, Chenyang window openers, GM40 curtain motor.

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

loongyh committed:
Initial commit (unfinished)

loongyh committed:
Second commit (unfinished)

loongyh committed:
Third commit (unfinished)

loongyh committed:
Fix compile errors

loongyh committed:
First working verison

loongyh committed:
Merge branch 'dev' into rs485

loongyh committed:
Linting fix

loongyh committed:
More linting fix

loongyh committed:
Add codeowners

OttoWinter committed:
Rewrite sun component calculations (#1661)

loongyh committed:
Add chenyang cover

loongyh committed:
Fix checksum bug, update codeowners

loongyh committed:
Fix checksum bug (2)

OttoWinter commented on esphome/components/rs485/rs485.cpp:
```diff
@@ -0,0 +1,74 @@
+  }
+}
+
+void RS485::parse_rs485_frame_() {
```
Wikipedia tells me RS485 just defines an electrical interface, not a communication protocol: https://en.wikipedia.org/wiki/RS-485#Protocols - so rs485 would be the wrong name for this component.

Is there some other specification that defines *the communication protocl*? We should use the name of that protocol for this component.

loongyh replied:
Hmm theoretically this is protocol-less and can be used for any multidrop/multipoint communication aside from RS485. It is named RS485 as currently I have a bunch of RS485 devices from different manufacturers sporting different protocols that I'd like to string up in a single bus controlled by a single UART on the ESP8266/32. Perhaps a better name may be "multidrop UART" or something?

OttoWinter commented on esphome/components/chenyang/chenyang.cpp:
```diff
@@ -0,0 +1,133 @@
+}
+
+void Chenyang::loop() {
+  if ((millis() - this->last_update_) > this->update_interval_) {
```
If this `loop()` is just here for scheduling recurring updates you should use `PollingComponent`

OttoWinter commented on esphome/components/rs485/rs485.cpp:
```diff
@@ -0,0 +1,74 @@
+
+void RS485::dump_config() {
+  ESP_LOGCONFIG(TAG, "RS485:");
+  ESP_LOGCONFIG(TAG, "%zu configured devices", this->devices_.size());
```
```suggestion
  ESP_LOGCONFIG(TAG, "%u configured devices", this->devices_.size());  // NOLINT
```

Unfortunately, ESP8266's printf does not support `%zu` - so instead use `%u` and tell clang tidy not to complain with NOLINT

OttoWinter reviewed (commented):
Thanks!

loongyh committed:
Change to PollingComponent

loongyh committed:
Run black formatter

OttoWinter commented on esphome/components/chenyang/cover.py:
```diff
@@ -0,0 +1,42 @@
+            ): cv.positive_time_period_milliseconds,
+        }
+    )
+    .extend(cv.COMPONENT_SCHEMA)
```
Instead of setting `update_interval` above, write:
```suggestion
    .extend(cv.polling_component_schema("1s"))
```

loongyh replied:
Isn't this hardcoding the update interval? Will the user still be able to configure it via `update_interval:` in .yaml?

loongyh replied:
`def polling_component_schema(default_update_interval):`

oh, just saw config_validation.py, yep it's still configurable: `Optional(CONF_UPDATE_INTERVAL, default=default_update_interval): update_interval,`

OttoWinter commented on esphome/components/chenyang/cover.py:
```diff
@@ -0,0 +1,42 @@
+        address = config[CONF_ADDRESS]
+        cg.add(var.set_address(address))
+
+    if CONF_UPDATE_INTERVAL in config:
```
With `cv.polling_component_schema` this is also no longer necessary (the update interval will be set in `cg.register_component`

OttoWinter commented on esphome/components/rs485/rs485.cpp:
```diff
@@ -0,0 +1,74 @@
+
+void RS485::loop() {
+  const uint32_t now = millis();
+  if (now - this->last_rs485_byte_ > 50) {
```
So if two devices (or even one) on the bus send messages close to each other, this will interpret that as one big message, right?

That sounds like it's just asking for trouble...

Most UART protocols include a start byte to synchronize the reading end - the dooya/chenyang should listen directly for that (like most other UART integrations in ESPHome). The multidrop uart integration could then just serve as a proxy buffering the received data and forwarding it to all registered listeners.

loongyh replied:
Devices on the bus are slaves and thus they can't initiate sending, it'll only respond upon receipt of a frame addressed to it. These frames are usually commands or status requests. Thus the 50ms is the timeout for the slave to send the response.

loongyh replied:
With these wildly different frame formats, even the start code may be different in Chenyang's case. To identify the component to forward a received frame to, in Dooya's case it's the combination of `start_code`, `address_h`, `address_l`, but for Chenyang's case we can't use the `start_code` as they differ, thus we can only use the `address` (2nd byte). Hence the first byte of Chenyang's `this->header_` is `nullptr` to signify ignore.

OttoWinter replied:
Correct, the header can be different for each device. What I mean is: each component (like chenyang) listens directly on the bus. It reads from the bus, waits for the start code and tries to parse the frame. if it's not a valid frame (or for another device), just ignore it. Or expressed another way: the chenyang component should be implemented just like the other esphome uart components, reading in stuff byte-by-byte.

How to do multidrop then? Have a new central component that reads from the uart bus, then sends the received bytes directly to all listening components like chenyang.

This way there's less latency (as we don't need to wait 50ms to ensure we're at the end of the frame), and the individual components are closer to ESPHome's UART abstractions.

loongyh replied:
So in Ethernet terms, instead of the central component acting as a switch, which routes packets only to the destined interface, we should have it act as a hub instead, which just blindly fires all packets to all interfaces. I was concerned with implementing a switch instead of a hub as it is more 'efficient', and totally overlooked the fact that the delay of 50ms actually makes it less efficient than just firing all received bytes to all registered components :facepalm:

loongyh replied:
If the base component should be multidrop `UART`, then perhaps we can modify `uart.h` to directly support multidrop devices. Meaning instead of implementing `void register_device(Device *device)` in a seperate `RS485` class, we put this directly in the `UARTDevice` class. Such that multiple integrations will be able to extend from the same `UARTDevice` instance, which if I understand correctly that'll mean they'll all have access to the same incoming and outgoing buffers.

OttoWinter replied:
Yeah implementing in the UARTDevice would of course be ideal. But that's probably not possible given what you mentioned. However, we could convert UARTComponent to an abstract class (which is implemented by "native" UART and multidrop); not sure if that would be so simple though because the multidrop interface would ideally use a callback to send each byte to the listeners (instead of having to children poll themselves)

loongyh replied:
I have reimplemented the RS485 base component into a `UARTMulti` component, with received bytes being passed to each component via `on_uart_multi_byte()`. This offloads parsing of the bytes-to-frames to the components themselves and affords a more 'native' access to the received bytes :).

loongyh committed:
Change to cv.polling_component_schema

loongyh committed:
Linting fix

OttoWinter committed:
Raise minimum python version to 3.7 (#1673)

OttoWinter committed:
Fix colorlog removing colors and refactor color code (#1671)

rbaron committed:
Adds support for b-parasite soil moisture sensor (#1666)

dataway committed:
mqtt_client: Added MQTTClientComponent::unsubscribe() (#1672)

Co-authored-by: Otto Winter <otto@otto-winter.com>

glmnet committed:
Fix servo detach chopped PWM (#1650)

pkuehne committed:
Disallow _ in node name (#1632)

gaco79 committed:
Receive long MQTT payload (#1590)

OttoWinter committed:
Fix sensor.sensor_schema interface changed (#1659)

dashdrum committed:
Daylight Saving Time spelling fix (#1677)

SenexCrenshaw committed:
Support custom build_flags for bme680_bsec (#1678)

OttoWinter committed:
Add Arduino ESP32 version mapping (#1679)

See also https://github.com/platformio/platform-espressif32/releases/tag/v3.2.0

OttoWinter committed:
Automate building and publishing of esphome-lint docker image (#1680)

SenexCrenshaw committed:
Sgp40 (#1513)

* Start of SGP40 dev

* Clean up

* Initial Commit

* VOC is working

* Fixed up sensor config

* Lint Fixes
Added in save/restore baseline
Noted original repo in header

* Lint Fixes
Added to test

* Lint Fixes

* Added additional check on restoring

* Removed double check

* Changed defines to static const double

* Changed defines to const
Do not send voc index until sensor stabilizes

* Fixed sensor stabilization message

* Fixup according to PR

* samples_read increment fix

* Fixed missing device class

* Choose a SENSOR device class

* Moved some sensors for tests

Co-authored-by: Guillermo Ruffino <glm.net@gmail.com>

dependabot[bot] committed:
Bump protobuf from 3.15.7 to 3.15.8 (#1682)

Bumps [protobuf](https://github.com/protocolbuffers/protobuf) from 3.15.7 to 3.15.8.
- [Release notes](https://github.com/protocolbuffers/protobuf/releases)
- [Changelog](https://github.com/protocolbuffers/protobuf/blob/master/generate_changelog.py)
- [Commits](https://github.com/protocolbuffers/protobuf/compare/v3.15.7...v3.15.8)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

OttoWinter committed:
Sensor Average Filter Fix Floating Pointer Error Accumulating (#1624)

coogle committed:
Implementing the remainder of GPS data for the GPS component. (#1676)

mmakaay committed:
Bump AsyncTCP-esphome to 1.2.1. (#1693)

Co-authored-by: Maurice Makaay <mmakaay1@xs4all.net>

richardklingler committed:
Added / to default glyphs (#1691)

Co-authored-by: Charlie Root <klingler@blender.klingler.net>

glmnet committed:
Revert "Bump AsyncTCP-esphome to 1.2.1. (#1693)" (#1709)

This reverts commit aed6f2b1ea860d9db73f53e99cd6982b38b5593c.

loongyh committed:
Refactor to UARTMulti

loongyh committed:
Linting fix

loongyh committed:
Fix missed commands

glmnet committed:
Change mac add mac suffix from underscore to dash (#1702)

loongyh committed:
GM40: Initial commit

ferbar committed:
Add monochromatic effects: Pulse, Random (#1616)

sjtrny committed:
Add support for SHT4X (#1512)

elyorkhakimov committed:
Addition of forward and reverse active energy counters to ATM90E32 sensor component (#1271)

Co-authored-by: Elyor Khakimov <elyorkhakimov@forwardnetworks.com>

loongyh committed:
Add Grow Fingerprint Reader (#1356)

glmnet committed:
RC522 fixes (#1479)

wstrzalka committed:
Support for TOF10120 distance sensor (#1375)

alex-richards committed:
Swap fan and swing fields for Fujitu ACs (#1635)

loongyh committed:
Invert pos

loongyh committed:
Change update() to callback. Invert GM40 set_pos.

loongyh committed:
Fix invalid syntax

loongyh committed:
GM40: Fix set_pos

loongyh committed:
Rate limit update callback. Fix GM40 set_pos.

loongyh committed:
send_update() only on ready_to_tx

loongyh committed:
GM40: Add ±3% error margin to idle state

loongyh committed:
GM40: Fix process_status()

loongyh committed:
GM40: Fix process_status()(2)

loongyh committed:
Linting fix

loongyh committed:
Merge branch 'dev' into rs485

loongyh committed:
Remove update_intervals in test3.yaml

loongyh commented:
The base `UARTMulti` component now cycles around each child component to trigger its update callback, with a default of 500ms interval. Tested fully working with the following deployments:
- 2 sets of (2x Chenyang window openers, 1x GM40 curtain motor)
- 2 sets of (1x Chenyang window openers, 1x GM40 curtain motor)
- 1 set of (1x Chenyang window opener, 1x Dooya shade motor)

loongyh committed:
Merge branch 'esphome:dev' into rs485

loongyh committed:
Linting fixes

loongyh committed:
Merge branch 'esphome:dev' into rs485

jesserockz commented:
I fail to see why these platforms specifically require the use of `UARTMulti`, based on the code of UARTMulti.

The way I see it is that `UARTMulti` should be an **optional** middleware component between `uart` and each platform so a user can just use a single device without the extra. And this would allow simple adaptation of other uart command/response devices to use the uart_multi bus...

loongyh commented:
That's true, I'll see to decoupling the platforms from uart_multi. Looking through the rest of the integrations with middleware, I don't see any optional ones, are there examples?
i.e. is there a way in config yaml to selectively choose to have the component extend the `UARTMultiDevice` or `UARTDevice` class?

jesserockz commented:
No it's not a thing at the moment...
Not sure exactly how it could be done either.

loongyh committed:
Merge branch 'dev' into rs485

loongyh committed:
Chenyang - Set pos to 0.5 if device uninitialized

loongyh committed:
Dooya, GM40 - Set pos to 0.5 if uninitialized
